File: frontend/src/services/courseService.js
import axios from 'axios';

const API_URL = 'http://localhost:8080/api/courses';

const getAuthHeader = () => {
  const token = localStorage.getItem('token');
  return token ? { Authorization: `Bearer ${token}` } : {};
};

const courseService = {
  getAllCourses: async () => {
    const response = await axios.get(API_URL, {
      headers: getAuthHeader()
    });
    return response.data;
  },

  getCourseById: async (id) => {
    const response = await axios.get(`${API_URL}/${id}`, {
      headers: getAuthHeader()
    });
    return response.data;
  },

  getCoursesByProgram: async (program) => {
    const response = await axios.get(`${API_URL}/program/${encodeURIComponent(program)}`, {
      headers: getAuthHeader()
    });
    return response.data;
  },

  getActiveCourses: async () => {
    const response = await axios.get(`${API_URL}/active`, {
      headers: getAuthHeader()
    });
    return response.data;
  },

  createCourse: async (courseData) => {
    const response = await axios.post(API_URL, courseData, {
      headers: getAuthHeader()
    });
    return response.data;
  },

  updateCourse: async (id, courseData) => {
    const response = await axios.put(`${API_URL}/${id}`, courseData, {
      headers: getAuthHeader()
    });
    return response.data;
  },

  deleteCourse: async (id) => {
    const response = await axios.delete(`${API_URL}/${id}`, {
      headers: getAuthHeader()
    });
    return response.data;
  }
};

export default courseService;

File: backend/src/main/java/com/flexisaf/admissions/service/CourseService.java
package com.flexisaf.admissions.service;

import com.flexisaf.admissions.entity.Course;
import com.flexisaf.admissions.exception.ResourceNotFoundException;
import com.flexisaf.admissions.repository.CourseRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Slf4j
public class CourseService {

    private final CourseRepository courseRepository;

    @Transactional(readOnly = true)
    public List<Course> getAllCourses() {
        log.info("Fetching all courses");
        return courseRepository.findAll();
    }

    @Transactional(readOnly = true)
    public Course getCourseById(Long id) {
        log.info("Fetching course with ID: {}", id);
        return courseRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Course", "id", id));
    }

    @Transactional(readOnly = true)
    public List<Course> getCoursesByProgram(String program) {
        log.info("Fetching courses for program: {}", program);
        return courseRepository.findByProgram(program);
    }

    @Transactional(readOnly = true)
    public Course getCourseByCode(String code) {
        log.info("Fetching course with code: {}", code);
        return courseRepository.findByCode(code)
                .orElseThrow(() -> new ResourceNotFoundException("Course", "code", code));
    }

    @Transactional(readOnly = true)
    public List<Course> getActiveCourses() {
        log.info("Fetching all active courses");
        return courseRepository.findByActiveTrue();
    }

    @Transactional
    public Course createCourse(Course course) {
        log.info("Creating new course: {}", course.getName());

        // Check if course code already exists
        courseRepository.findByCode(course.getCode()).ifPresent(existing -> {
            throw new IllegalArgumentException("Course with code " + course.getCode() + " already exists");
        });

        Course savedCourse = courseRepository.save(course);
        log.info("Successfully created course with ID: {}", savedCourse.getId());
        return savedCourse;
    }

    @Transactional
    public Course updateCourse(Long id, Course updatedCourse) {
        log.info("Updating course with ID: {}", id);

        Course existingCourse = courseRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Course", "id", id));

        // Check if course code is being changed and if new code already exists
        if (!existingCourse.getCode().equals(updatedCourse.getCode())) {
            courseRepository.findByCode(updatedCourse.getCode()).ifPresent(existing -> {
                throw new IllegalArgumentException("Course with code " + updatedCourse.getCode() + " already exists");
            });
        }

        // Update fields
        existingCourse.setCode(updatedCourse.getCode());
        existingCourse.setName(updatedCourse.getName());
        existingCourse.setCredits(updatedCourse.getCredits());
        existingCourse.setInstructor(updatedCourse.getInstructor());
        existingCourse.setSchedule(updatedCourse.getSchedule());
        existingCourse.setProgram(updatedCourse.getProgram());
        existingCourse.setDescription(updatedCourse.getDescription());
        existingCourse.setCapacity(updatedCourse.getCapacity());
        existingCourse.setEnrolled(updatedCourse.getEnrolled());
        existingCourse.setActive(updatedCourse.getActive());

        Course saved = courseRepository.save(existingCourse);
        log.info("Successfully updated course with ID: {}", id);
        return saved;
    }

    @Transactional
    public void deleteCourse(Long id) {
        log.info("Deleting course with ID: {}", id);

        Course course = courseRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Course", "id", id));

        courseRepository.delete(course);
        log.info("Successfully deleted course with ID: {}", id);
    }
}
